gadgets/traceloop/program.bpf.c

```diff
@@ -0,0 +1,573 @@
+	/*
+	 * We have around 300 syscalls, let's use the immediate greater power of
+	 * 2.
+	 */
+	__uint(max_entries, 512);
+} syscalls SEC(".maps");
+
+/*
+ * This key/value store maps thread PIDs to syscall arg arrays
+ * that were remembered at sys_enter so that sys_exit can probe buffer
+ * contents and generate syscall events showing the result content.
+ */
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(key_size, sizeof(u64));
+	__uint(value_size, sizeof(struct remembered_args));
+	__uint(max_entries, 1024);
+} probe_at_sys_exit SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(key_size, sizeof(u64));
+	__uint(value_size, sizeof(struct pt_regs));
+	__uint(max_entries, 1024);
+} regs_map SEC(".maps");
+
+static __always_inline int skip_exit_probe(int nr)
+{
+	return !!(nr == __NR_exit || nr == __NR_exit_group ||
+		  nr == __NR_rt_sigreturn);
+}
+
+/*
+ * Highly inspired from ksnoop.bpf.c:
+ * https://github.com/iovisor/bcc/blob/f90126bb3770ea1bdd915ff3b47e451c6dde5c40/libbpf-tools/ksnoop.bpf.c#L280
+ */
+static __always_inline u64 get_arg(struct pt_regs *regs, int i)
+{
+	switch (i) {
+	case 1:
+		return PT_REGS_PARM1_CORE_SYSCALL(regs);
+	case 2:
+		return PT_REGS_PARM2_CORE_SYSCALL(regs);
+	case 3:
+		return PT_REGS_PARM3_CORE_SYSCALL(regs);
+	case 4:
+		return PT_REGS_PARM4_CORE_SYSCALL(regs);
+	case 5:
+		return PT_REGS_PARM5_CORE_SYSCALL(regs);
+	case 6:
+		return PT_REGS_PARM6_CORE_SYSCALL(regs);
+	default:
+		bpf_error_printk(
+			"There is no PT_REGS_PARM%d_SYSCALL macro, check the argument!",
+			i);
+		return 0;
+	}
+}
+
+/*
+ * sys_enter is defined as:
+ * TP_PROTO(struct pt_regs *regs, long id)
+ * (https://elixir.bootlin.com/linux/v5.19/source/include/trace/events/syscalls.h#L20)
+ * So, ctx->args[0] contains a struct pt_regs and ctx->args[1] the syscall ID.
+ */
+SEC("raw_tracepoint/sys_enter")
+int ig_traceloop_e(struct bpf_raw_tracepoint_args *ctx)
+{
+	struct remembered_args remembered = {};
+	u64 pid = bpf_get_current_pid_tgid();
+	struct syscall_def_t *syscall_def;
+	/*
+	 * Initialize struct to empty to be sure all fields (even padding) are zeroed:
+	 * https://github.com/iovisor/bcc/issues/2623#issuecomment-560214481
+	 */
+	struct syscall_event_t sc = {};
+	struct task_struct *task;
+	u64 nr = ctx->args[1];
+	struct pt_regs *args;
+	void *perf_buffer;
+	u64 mntns_id;
+	int ret;
+	int i;
+
+	/* The boot time timestamp is used to give the timestamp to users. It
+	 * is converted to the wall-clock time in userspace. It only works
+	 * from Linux 5.7. On older kernels, the BPF bytecode for
+	 * bpf_ktime_get_boot_ns is automatically removed by the BPF loader,
+	 * see FixBpfKtimeGetBootNs. In this way, this BPF program can still be
+	 * loaded on older kernels. */
+	u64 boot_ts = bpf_ktime_get_boot_ns();
+
+	/* The monotonic timestamp is used by traceloop to match the sys_enter
+	 * event with the cont and sys_exit events. This is an internal
+	 * implementation detail not exposed to the user. */
```

Comment:
Is it possible that two syscalls get executed at the exact same time on different cpus? Should the cpu number be added to this key as well?